crates/fluvio/src/consumer/retry.rs

```
@@ -0,0 +1,279 @@
+    async fn consumer_with_retry(
+        inner: ConsumerWithRetryInner,
+        fluvio_client: Arc<Fluvio>,
+        mut current_stream: BoxConsumerStream,
+    ) -> Option<Result<(ConsumerRecord, BoxConsumerStream, Option<i64>), ErrorCode>> {
+        let mut attempts: u32 = 0;
+        let mut backoff = create_backoff();
+
+        loop {
+            // Try to retrieve the next record.
+            if let Some(record_result) = current_stream.as_mut().next().await {
+                match record_result {
+                    Ok(record) => {
+                        let new_offset = Some(record.offset + 1);
+                        if attempts > 0 {
+                            debug!(
+                                target: SPAN_RETRY,
+                                "Record produced successfully after reconnect"
+                            );
+                        }
+                        return Some(Ok((record, current_stream, new_offset)));
+                    }
+                    Err(e) => {
+                        warn!(target: SPAN_RETRY, "Error consuming record: {}", e);
+                        if let RetryMode::Disabled = inner.consumer_config.retry_mode {
+                            return Some(Err(e));
+                        }
+                    }
+                }
+            }
+
+            // If continuous consumption is disabled, end the stream.
+            if inner.consumer_config.disable_continuous {
+                return None;
+            }
+
+            // Wait before retrying.
+            backoff_and_wait(&mut backoff).await;
+            attempts += 1;
+
+            // Determine the offset for reconnection.
+            let offset = if let Some(next) = inner.next_offset_to_read {
+                match Offset::absolute(next) {
+                    Ok(off) => off,
+                    Err(e) => {
+                        warn!(target: SPAN_RETRY, "Error creating offset: {}", e);
```

return error?
or alternatively, have option to ignore error

---

We're returning error at next line, no?

---

k